modules/backendscheduler/backendscheduler.go

```diff
@@ -0,0 +1,391 @@
+	}
+
+	return j, nil
+}
+
+// CompleteJob marks a job as completed
+func (s *BackendScheduler) CompleteJob(_ context.Context, id string) error {
+	j := s.work.GetJob(id)
+	if j == nil {
+		return fmt.Errorf("job not found: %s", id)
+	}
+
+	j.Complete()
+	metricJobsCompleted.WithLabelValues(j.JobDetail.Tenant, j.Type.String()).Inc()
+	metricJobsActive.WithLabelValues(j.JobDetail.Tenant, j.Type.String()).Dec()
+
+	level.Info(log.Logger).Log("msg", "completed job", "job_id", id, "tenant", j.JobDetail.Tenant, "type", j.Type.String())
+	return nil
+}
+
+// FailJob marks a job as failed
+func (s *BackendScheduler) FailJob(_ context.Context, id string) error {
+	j := s.work.GetJob(id)
+	if j == nil {
+		return fmt.Errorf("job not found: %s", id)
+	}
+
+	j.Fail()
+	metricJobsFailed.WithLabelValues(j.JobDetail.Tenant, j.Type.String()).Inc()
+	metricJobsActive.WithLabelValues(j.JobDetail.Tenant, j.Type.String()).Dec()
+
+	level.Info(log.Logger).Log("msg", "failed job", "job_id", id, "tenant", j.JobDetail.Tenant, "type", j.Type.String())
+	return nil
+}
+
+// Next implements the BackendSchedulerServer interface
+func (s *BackendScheduler) Next(_ context.Context, req *tempopb.NextJobRequest) (*tempopb.NextJobResponse, error) {
+	// Find jobs that already exist for this worker
+	for id, j := range s.work.Jobs() {
+		if j.Status() == work.JobStatusPending || j.Status() == work.JobStatusRunning {
+			if j.WorkerID() == req.WorkerId {
+				resp := &tempopb.NextJobResponse{
+					JobId:  j.ID,
+					Type:   j.Type,
+					Detail: j.JobDetail,
+				}
+
+				level.Info(log.Logger).Log("msg", "assigned previous job to worker", "job_id", id, "worker", req.WorkerId)
+				return resp, nil
+			}
+		}
+	}
+
+	// Find next available job
+	for id, j := range s.work.Jobs() {
+		if j.Status() == work.JobStatusPending {
+			// Honor the request job type if specified
+			if req.Type != tempopb.JobType_JOB_TYPE_UNSPECIFIED && j.Type != req.Type {
+				continue
+			}
+
+			// Create response with job details
+			resp := &tempopb.NextJobResponse{
+				JobId:  j.ID,
+				Type:   j.Type,
+				Detail: j.JobDetail,
+			}
+
+			// Mark job as running
+			j.Start(req.WorkerId)
+
+			level.Info(log.Logger).Log("msg", "assigned job to worker", "job_id", id, "worker", req.WorkerId)
+			return resp, nil
+		}
+	}
+
+	// No jobs available
+	return &tempopb.NextJobResponse{}, nil
+}
+
+// UpdateJob implements the BackendSchedulerServer interface
+func (s *BackendScheduler) UpdateJob(_ context.Context, req *tempopb.UpdateJobStatusRequest) (*tempopb.UpdateJobStatusResponse, error) {
+	j := s.work.GetJob(req.JobId)
+	if j == nil {
+		return nil, fmt.Errorf("job not found: %s", req.JobId)
+	}
+
+	switch req.Status {
+	case tempopb.JobStatus_JOB_STATUS_SUCCEEDED:
+		j.Complete()
+		metricJobsCompleted.WithLabelValues(j.JobDetail.Tenant, j.JobDetail.Tenant).Inc()
+		level.Info(log.Logger).Log("msg", "job completed", "job_id", req.JobId)
+
+		// TODO: update the blocklist?
```

Comment:
we may have to do this? pretty sure the current implementation does this to prevent constantly picking jobs that don't actually exist. good catch.

Reply:
This has been updated.